File: typescript/day01.ts
import { readFileSync } from 'fs';

const inputFilename: string = '../data/input01.txt';

type Calorie = number;
type Elf = Calorie[];
type Elves = Elf[];
type Partition = [Elf, Elves];

const sum = (ns: number[]) => ns.reduce((sum, n) => sum + n, 0);

const parse = (lines: string[]): Elves =>
  lines.reduce(
    ([elf, elves], line): Partition => {
      if (line.trim() === '') {
        elf.length > 0 && elves.push(elf);
        return [[], elves];
      } else {
        const calories = parseInt(line);
        elf.push(calories);
        return [elf, elves];
      }
    },
    [[], []] as Partition
  )[1];

const text = readFileSync(inputFilename, 'utf8').split('\n');
const elves: Elves = parse(text);
const calories = elves.map(sum).sort((a, b) => b - a);

const [top] = calories;
const topThree = sum(calories.slice(0, 3));

console.log({ top, topThree });

File: haskell/app/Day01.hs
module Day01 (main) where

import Helpers (yoloReadFile)
import Data.List (sort)

inputData :: String
inputData = yoloReadFile "../data/input01.txt"

type Calorie = Integer
type Elf = [Calorie]
type Elves = [Elf]

-- parse lines into groups of integers
parse :: [String] -> Elves
parse strings = parse' strings [] []

parse' :: [String] -> Elf -> Elves -> Elves
parse' [] elf elves = reverse (elf:elves)
parse' ("":xs) [] elves = parse' xs [] elves
parse' ("":xs) elf elves = parse' xs [] (reverse elf:elves)
parse' (x:xs) elf elves = parse' xs (read x:elf) elves


main :: IO ()
main = print (top, topThree)
  where elves = parse (lines inputData)
        sums = reverse (sort (map sum elves))
        top = head sums
        topThree = sum (take 3 sums)
